### PR TITLE
Added typescript generator

### PR DESCRIPTION
**Pull request by rg911**
### Typescript Generator

Typescript generator created in this PR:
- **Typescript**: Strong typed typescript generator.

some examples generated code below

#### Typescript Generator
```js

import { GeneratorUtils } from './GeneratorUtils';
import { AccountLinkActionDto } from './AccountLinkActionDto';
import { EntityTypeDto } from './EntityTypeDto';
import { SignatureDto } from './SignatureDto';
import { AccountLinkTransactionBodyBuilder } from './AccountLinkTransactionBodyBuilder';
import { TimestampDto } from './TimestampDto';
import { KeyDto } from './KeyDto';
import { AmountDto } from './AmountDto';
import { TransactionBuilder } from './TransactionBuilder';

/** Binary layout for a non-embedded account link transaction. */
export class AccountLinkTransactionBuilder extends TransactionBuilder {
    /** Account link transaction body. */
    accountLinkTransactionBody: AccountLinkTransactionBodyBuilder;

    /**
     * Constructor.
     *
     * @param signature Entity signature.
     * @param signer Entity signer's public key.
     * @param version Entity version.
     * @param type Entity type.
     * @param fee Transaction fee.
     * @param deadline Transaction deadline.
     * @param remoteAccountKey Remote account key.
     * @param linkAction Account link action.
     */
    public constructor(signature: SignatureDto,  signer: KeyDto,  version: number,  type: EntityTypeDto,  fee: AmountDto,  deadline: TimestampDto,  remoteAccountKey: KeyDto,  linkAction: AccountLinkActionDto) {
        super(signature, signer, version, type, fee, deadline);
        this.accountLinkTransactionBody = AccountLinkTransactionBodyBuilder.create(remoteAccountKey, linkAction);
    }

    /**
     * Gets remote account key.
     *
     * @return Remote account key.
     */
    public getRemoteAccountKey(): KeyDto {
        return this.accountLinkTransactionBody.getRemoteAccountKey();
    }

    /**
     * Gets account link action.
     *
     * @return Account link action.
     */
    public getLinkAction(): AccountLinkActionDto {
        return this.accountLinkTransactionBody.getLinkAction();
    }

    /**
     * Gets the size of the object.
     *
     * @return Size in bytes.
     */
    public getSize(): number {
        let size: number = super.getSize();
        size += this.accountLinkTransactionBody.getSize();
        return size;
    }

    /**
     * Creates an instance of AccountLinkTransactionBuilder from binary payload.
     *
     * @param payload Byte payload to use to serialize the object.
     * @return Instance of AccountLinkTransactionBuilder.
     */
    public static loadFromBinary(payload: Uint8Array): AccountLinkTransactionBuilder {
        let byteArray = Array.from(payload);
        const superObject = TransactionBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        const accountLinkTransactionBody = AccountLinkTransactionBodyBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, accountLinkTransactionBody.getSize());
        return new AccountLinkTransactionBuilder(superObject.signature, superObject.signer, superObject.version, superObject.type, superObject.fee, superObject.deadline, accountLinkTransactionBody.remoteAccountKey, accountLinkTransactionBody.linkAction);
    }

    /**
     * Serializes an object to bytes.
     *
     * @return Serialized bytes.
     */
    public serialize(): Uint8Array {
        let newArray = Uint8Array.from([]);
        const superBytes = super.serialize();
        newArray = GeneratorUtils.concatTypedArrays(newArray, superBytes);
        const accountLinkTransactionBodyBytes = this.accountLinkTransactionBody.serialize();
        newArray = GeneratorUtils.concatTypedArrays(newArray, accountLinkTransactionBodyBytes);
        return newArray;
    }
}
```

### CODE REVIEW

File: generators/typescript/GeneratorUtils.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
+/**
+ * Generator utility class.
+ */
+export class GeneratorUtils {
+
+	/**
+     * Converts a (64bit) uint8 array into a number array.
```
Comment (evias) [inline generators/typescript/GeneratorUtils.ts]: `Convert a UInt8Array input into an array of 2 numbers. Numbers in the returned array are cast to UInt32.`

File: generators/typescript/GeneratorUtils.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    public static bufferToUint64 (input: Uint8Array): number[] {
+        if (8 !== input.length) {
+            throw Error(`byte array has unexpected size '${input.length}'`);
+		}
+		input = input.reverse();
+        return [GeneratorUtils.readUint32At(input, 0), GeneratorUtils.readUint32At(input, 4)];
+	}
+	
+	/**
+     * Read buffer into 32bits integer at given index.
```
Comment (evias) [inline generators/typescript/GeneratorUtils.ts]: "Read 4 bytes as a uint32 value from buffer `bytes` starting at given index."

File: generators/typescript/GeneratorUtils.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+	/**
+     * Write uint to buffer
+     * @param {number} uintValue A uint8 array.
+	 * @param {number} bufferSize Buffer size.
+     * @returns {Uint8Array}
+     */
+	public static uintToBuffer (uintValue: number, bufferSize: number): Uint8Array {
+		const buffer = new ArrayBuffer(bufferSize);
+		const dataView = new DataView(buffer);
+		if (1 === bufferSize)
```
Comment (evias) [inline generators/typescript/GeneratorUtils.ts]: does it make sense to add an overflow validation ? (uint8 max = 255)

File: generators/typescript/GeneratorUtils.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+			dataView.setUint8(0, uintValue);
+	
+		else if (2 === bufferSize)
```
Comment (evias) [inline generators/typescript/GeneratorUtils.ts]: does it make sense to add an overflow validation ? (uint16 max = 65535)

File: generators/typescript/GeneratorUtils.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			dataView.setUint32(0, uintValue, true);
+	
+		else
+			throw new Error('Unexpected bufferSize');
+	
+		return new Uint8Array(buffer);
+	}
+
+	/**
+     * Write uint to buffer
```
Comment (evias) [inline generators/typescript/GeneratorUtils.ts]: `Convert uint8 array buffer into number`

File: generators/typescript/TypescriptClassGenerator.py [hunk 6]
```diff
@@ -0,0 +1,725 @@
...
+        method = '{0}(this.{1}(), {2})'.format(get_read_method_name(size),
+                                               self._get_generated_getter_name(attribute['name']),
+                                               size) if size < 8 else '{0}(this.{1}())'.format(get_read_method_name(size),
+                                                                                               self._get_generated_getter_name(
+                                                                                                   attribute['name']))
+        lines = ['const {0}Bytes = {1}'.format(attribute['name'], method)]
+        lines += ['newArray = GeneratorUtils.concatTypedArrays(newArray, {0})'.format(attribute['name']+'Bytes')]
+        self._add_attribute_condition_if_needed(attribute, serialize_method, 'this.', lines, False)
+
+    def _serialize_attribute_array(self, attribute, serialize_method):
```
Comment (Wayonb) [inline generators/typescript/TypescriptClassGenerator.py]: I think you will need to add support for the new __FILL__ attribute if you want aggregate tx to work.  I have not updated the Java either.

File: generators/typescript/TypescriptFileGenerator.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    def set_import(self):
+        # self.code += ['']
+        pass
+
+    def _initialize_class(self):
+        self.code = []
+        self.add_copyright(self.options['copyright'])
+
+    def generate(self):
+        for type_descriptor, value in self.schema.items():
```
Comment (Wayonb) [inline generators/typescript/TypescriptFileGenerator.py]: Is this code similar to Java except for the file extension been appended?

File: generators/typescript/TypescriptStaticClassGenerator.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
+# pylint: disable=too-few-public-methods
+from inspect import getframeinfo, currentframe
+from os.path import dirname, abspath, realpath, join
+
+
+class TypescriptStaticClassGenerator():
```
Comment (Wayonb) [inline generators/typescript/TypescriptStaticClassGenerator.py]: Same for this file.  Can it be just a common helper between the generators?

File: generators/typescript/Helpers.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
+from enum import Enum
+
+
+class TypeDescriptorDisposition(Enum):
+    Inline = 'inline'
+    Const = 'const'
+
+
+def indent(code, n_indents=1):
```
[the remark this answers is not in the document]
Comment (rg911) [reply generators/typescript/Helpers.py]: Agreed with Wayonb that we keep the generator code separate per language

File: generators/typescript/TypescriptFileGenerator.py (hunk 7 above)
Comment (rg911) [reply generators/typescript/TypescriptFileGenerator.py]: Agreed with Wayonb that we keep the generator code separate per language

File: generators/typescript/TypescriptStaticClassGenerator.py (hunk 8 above)
Comment (rg911) [reply generators/typescript/TypescriptStaticClassGenerator.py]: Agreed with Wayonb that we keep the generator code separate per language

**Comment by Wayonb** [comment]
The patch that you added for aggregate tx will work but the transactions and cosignatures are just byte array instead to the dto type. 

transactions: Uint8Array ==> TransactionDto
cosignatures: Uint8Array ==> CosignatureDto

File: generators/typescript/TypescriptClassGenerator.py [hunk 10]
```diff
@@ -286,6 +288,14 @@ def _recursive_attribute_iterator(self, class_name, callback, context, ignore_in
                             self.generated_class_name, attribute_val['value'],
                             get_comment_from_name(self.generated_class_name))
                     continue
+                elif attribute_val['disposition'] == TypeDescriptorDisposition.Var.value:
+                    attribute_val['type'] = 'byte'
```
Comment (Wayonb) [inline generators/typescript/TypescriptClassGenerator.py]: Should use the default type

File: generators/typescript/TypescriptClassGenerator.py [hunk 11]
```diff
@@ -286,6 +288,14 @@ def _recursive_attribute_iterator(self, class_name, callback, context, ignore_in
...
+                    callback(attribute_val, context)
+                    continue
+                elif attribute_val['disposition'] == TypeDescriptorDisposition.Fill.value:
+                    attribute_val['type'] = 'array'
```
Comment (Wayonb) [inline generators/typescript/TypescriptClassGenerator.py]: Same here.

File: generators/typescript/TypescriptClassGenerator.py (hunk 10 above)
Comment (rg911) [reply generators/typescript/TypescriptClassGenerator.py]: hmm. thought about it, looks like it is easier for ts to cope directly with buffer?

Comment (rg911) [reply generators/typescript/TypescriptClassGenerator.py]: sure mate, will get these cracked

File: generators/typescript/TypescriptClassGenerator.py (hunk 6 above)
Comment (fboucquez) [reply generators/typescript/TypescriptClassGenerator.py]: Both Java and TS Aggregate Builders need to be fixed. But we are currently releasing the SDKs with the branched version of catbuffer-generators. I think this improvement needs to be logged separately. It's not a trivial fix as we required the big switch statement for the different type of inner transactions. EntityType/TransactionType enum needs to be fixed first (at least in java)

**Comment by rg911** [comment]
Open a new PR

**Pull request closed without merging**